src/lib/replication/repl_dev/raft_repl_dev.cpp

```diff
@@ -0,0 +1,451 @@
+                                              });
+        }
+    } else {
+        m_data_journal = std::make_shared< ReplLogStore >(*this, *m_state_machine);
+        m_rd_sb->data_journal_id = m_data_journal->logstore_id();
+        m_rd_sb->last_applied_dsn = 0;
+        m_rd_sb->group_ordinal = s_next_group_ordinal.fetch_add(1);
+        m_rdev_name = fmt::format("rdev{}", m_rd_sb->group_ordinal);
+
+        if (m_rd_sb->is_timeline_consistent) {
+            m_free_blks_journal =
+                logstore_service().create_new_log_store(LogStoreService::CTRL_LOG_FAMILY_IDX, false /* append_mode */);
+            m_rd_sb->free_blks_journal_id = m_free_blks_journal->get_store_id();
+        }
+    }
+
+    RD_LOG(INFO, "Started {} RaftReplDev group_id={}, replica_id={}, raft_server_id={} commited_lsn={} next_dsn={}",
+           (load_existing ? "Existing" : "New"), group_id_str(), my_replica_id_str(), m_raft_server_id,
+           m_commit_upto_lsn.load(), m_next_dsn.load());
+    m_msg_mgr.bind_data_service_request(PUSH_DATA, m_group_id, bind_this(RaftReplDev::on_push_data_received, 1));
+    // m_msg_mgr.bind_data_service_request(FETCH_DATA, m_group_id, bind_this(RaftReplDev::on_fetch_data_received, 2));
+}
+
+void RaftReplDev::use_config(json_superblk raft_config_sb) { m_raft_config_sb = std::move(raft_config_sb); }
+
+void RaftReplDev::async_alloc_write(sisl::blob const& header, sisl::blob const& key, sisl::sg_list const& value,
+                                    repl_req_ptr_t rreq) {
+    if (!rreq) { auto rreq = repl_req_ptr_t(new repl_req_ctx{}); }
+    rreq->header = header;
+    rreq->key = key;
+    rreq->value = std::move(value);
+
+    // If it is header only entry, directly propose to the raft
+    if (rreq->value.size) {
+        rreq->rkey =
+            repl_key{.server_id = server_id(), .term = raft_server()->get_term(), .dsn = m_next_dsn.fetch_add(1)};
+        push_data_to_all_followers(rreq);
+
+        // Step 1: Alloc Blkid
+        auto status = data_service().alloc_blks(uint32_cast(rreq->value.size),
+                                                m_listener->get_blk_alloc_hints(rreq->header, rreq->value.size),
+                                                rreq->local_blkid);
+        HS_REL_ASSERT_EQ(status, BlkAllocStatus::SUCCESS);
+
+        // Write the data
+        data_service().async_write(rreq->value, rreq->local_blkid).thenValue([this, rreq](auto&& err) {
+            HS_REL_ASSERT(!err, "Error in writing data"); // TODO: Find a way to return error to the Listener
+            rreq->state.fetch_or(uint32_cast(repl_req_state_t::DATA_WRITTEN));
+            m_state_machine->propose_to_raft(std::move(rreq));
+            // if (rreq->state.load() & uint32_cast(repl_req_state_t::LOG_FLUSHED)) { report_committed(std::move(rreq));
+            // }
+        });
+    } else {
+        RD_LOG(INFO, "Skipping data channel send since value size is 0");
```

Comment:
we need to support the case without value, right? so we still need to call m_state_machine->propose_to_raft() in here

Reply:
Yes, modified the code.